Keys/ContainsKey/Values implementation with MergedDictionary

andiblas opened the pull request:
older implementation of these methods did not take into account the
_mergedInstance attribute, so the values returned did not reflect the
real dictionaries within.

dnfclas commented:
Hi __@andiblas__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla2.dotnetfoundation.org.
    </span>

TTYL, DNFBOT;

dnfclas commented:
__@andiblas__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

jassmith commented:
Could you provide a sample situation where this is needed?

andiblas commented:
I was tryning to use the new API MergedWith for ResourceDictionary. So to test this before applying it into our production app, i've created a new project, and added 2 ResourceDictionaries, let's say Parent.xaml and Child.xaml. In the child ResourceDictionary, i've added the MergedWith attribute, where i specify the Parent dictionary. Done this, i expect that the child dictionary has overriden items from the parent with the same key + it's own items. 

In the page i defined i'm using styles defined both in Child and Parent ResourceDictionaries. But here is where the trouble comes on: app crashes at InitializeComponent() of that page i'm trying to run. Digging a bit further, searching in the stacktrace i've found that the Exception comes from here:


[StaticResourceExtension.cs](https://github.com/xamarin/Xamarin.Forms/blob/2d9288eee6e6f197364a64308183725e7bd561f9/Xamarin.Forms.Xaml/MarkupExtensions/StaticResourceExtension.cs#L47)


The StaticResourceExtension class calls ContainsKey of ResourceDictionary, where i've found that doesn't take into account the merged dictionary within. The [indexer](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Core/ResourceDictionary.cs#L99) or [Count](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Core/ResourceDictionary.cs#L68) property actually read the values stored in _mergedInstance. So i think that this is the root cause of the issue, other properties or methods (ContainsKey, Keys or Values) are not exposing the real amount of keys the ResourceDictionary has. 

Hope i've made myself clear.

StephaneDelcroix commented:
@andiblas Please add an unittest

StephaneDelcroix commented:
@andiblas We really need a sample case, or better a unit test, showcasing the problem you are tying to solve.
Without that, we'll unfortunately have to close this PR...

StephaneDelcroix commented:
Is this the bug you are trying to fix https://bugzilla.xamarin.com/show_bug.cgi?id=42264 ?

StephaneDelcroix commented:
while trying to understand your issue, I came up with a different set of fixes in #536. closing this one
